lib/view_component/i18n.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require "yaml"
+require "pathname"
+
+module ViewComponent
+  module I18n
+    # Expects a component to have a file with the same basename and
+    # the .yml extension, holding translations scoped to the component's path.
+    #
+    # A component YML should look like this:
+    #
+    #   en:
+    #     hello: "Hello World!"
+    #
+    # And inside the component will be possible to call
+    #
+    #   t(".hello") # => "Hello World!"
+    #
+    # Internally the scope will be expanded using the template's path.
+    # As an example, the translation above from within `app/components/greeting/component.html.erb`
+    # will result in an internal I18n key like this:
+    #
+    #   en:
+    #     greeting:
+    #       component:
+    #         hello: "Hello World!"
```

Comment:
I'm curious about the rationale behind this naming/namespacing.

By default the naming of the template file for an `ExampleComponent` with sidecar would be: `app/components/example_component/example_component.html.erb` so I think the docs should show this case instead.

What would the structure of the I18n key be in this case? `en.example_component.example_component.hello`, right?

Reply:
:wave: @elia did you see this comment?

Reply:
_Sorry, I did read it but then somehow skipped it when replying_

I'll update the doc, the use of `example/component.*` was just coming from the way we organized them in the app I extracted this from.

Reply:
*Don't worry about it ;)*

I find the `en.example_component.example_component.hello` I18n key not ideal, and I'm wondering if it shouldn't be something like `en.view_component.example_component.hello` or some other naming that:

- uses a `view_component` or `components` namespace to prevent collision with other parts of the app
- removes the component name duplication
- is consistent whether you have:
    - a *regular* component (with a template placed in `app/components/example_component.html.erb`)
    - a *sidecar* component (with a template placed in `app/components/example_component/example_component.html.erb`)
    - an *inline* component (so the `I18n.t` method is called from the component `call` method)

What do you think?

Reply:
I like the idea of both the prefix and collapsing the sidecar components path with the duplicate name, the only thing I want to point out before trying to implement them is that we need to keep this configuration in synch with `virtual_path` in `ViewComponent::Base` because that's the one that will drive relative translations inside templates.

So, probably we should start moving some of this stuff to engine-level configurations, including maybe the components root and the glob for the YAML files. That would be ok?

Reply:
I'd like to request @joelhawksley and @BlakeWilliams opinion about this, as I'm not sure what's the best move.

Reply:
Building on https://github.com/github/view_component/pull/552#discussion_r584988143, if we go down that path I think we could avoid coupling i18n key names to specific components because each component will have access to the global i81n, but also its sidecar i18n, but not the sidecar i18n of other components. 

> So, probably we should start moving some of this stuff to engine-level configurations, including maybe the components root and the glob for the YAML files. That would be ok?

We should avoid configuration wherever possible (or at least have some great defaults), so if we can make i18n work out of the box that would be preferable.

Reply:
@BlakeWilliams please checkout #660 which features a version of sidecar i18n as you intended, the global one is still available via `I18n.t` or `helpers.t`.